File: clients/go-sdk/client/permissions.go
package client

import "context"

// Permissions are immutable platform constants; the only operations are
// list + get. Mutation of permission grants happens via the role itself
// — see RolesResource.GrantPermission / RevokePermission.

// PermissionResponse is one permission catalogue entry.
type PermissionResponse struct {
	Permission  string `json:"permission"`
	Application string `json:"application"`
	Context     string `json:"context"`
	Aggregate   string `json:"aggregate"`
	Action      string `json:"action"`
	Description string `json:"description"`
}

// PermissionListResponse — GET /api/roles/permissions.
type PermissionListResponse struct {
	Permissions []PermissionResponse `json:"permissions"`
	Total       uint64               `json:"total,omitempty"`
}

// PermissionsResource — /api/roles/permissions/*.
type PermissionsResource struct {
	c *FlowCatalystClient
}

// List — GET /api/roles/permissions.
func (r *PermissionsResource) List(ctx context.Context) (*PermissionListResponse, error) {
	var out PermissionListResponse
	if err := r.c.Get(ctx, "/api/roles/permissions", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/roles/permissions/{name}.
func (r *PermissionsResource) Get(ctx context.Context, name string) (*PermissionResponse, error) {
	var out PermissionResponse
	if err := r.c.Get(ctx, "/api/roles/permissions/"+name, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/permissions.rs
//! Permission catalogue queries — `/api/roles/permissions/*`.
//!
//! Permissions are immutable platform constants; the only operations are
//! list + get. Mutation of permission grants happens via the role itself
//! (see [`crate::client::Roles::grant_permission`]).

use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// Paginated list of permissions — `GET /api/roles/permissions`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PermissionListResponse {
    pub permissions: Vec<PermissionResponse>,
    #[serde(default)]
    pub total: u64,
}

/// Permission response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PermissionResponse {
    pub permission: String,
    pub application: String,
    pub context: String,
    pub aggregate: String,
    pub action: String,
    pub description: String,
}

/// Permissions resource accessor — created via [`FlowCatalystClient::permissions`].
pub struct Permissions<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl Permissions<'_> {
    /// List all permissions.
    pub async fn list(&self) -> Result<PermissionListResponse, ClientError> {
        self.client.get("/api/roles/permissions").await
    }

    /// Get a permission by name.
    pub async fn get(&self, name: &str) -> Result<PermissionResponse, ClientError> {
        self.client
            .get(&format!("/api/roles/permissions/{}", name))
            .await
    }
}
